py/dml/codegen_test.py

```diff
@@ -0,0 +1,110 @@
+# © 2021 Intel Corporation
+# SPDX-License-Identifier: MPL-2.0
+
+import unittest
+import math
+from dml.codegen import loss_on_truncation
+
+class Test_loss_on_truncation(unittest.TestCase):
+    def int_edge_testing(self, signed, max_bits=64, val_wrapper=lambda x: x):
+        def has_loss(v, bits):
+            self.assertTrue(loss_on_truncation(val_wrapper(v), bits, signed))
+        def no_loss(v, bits):
+            self.assertFalse(loss_on_truncation(val_wrapper(v), bits, signed))
+
+        for bits in range(2, max_bits+1):
+            for i in range(-1, 2):
+                no_loss(i, bits)
+            max_pos = 2**(bits-signed) - 1
+            # Unsigned are treated as groups; min neg allowed for those
+            # are -(2^bits - 1)
+            min_neg = -(2**(bits-signed) - (not signed))
+            no_loss(max_pos, bits)
+            no_loss(min_neg, bits)
+            has_loss(max_pos + 1, bits)
+            has_loss(min_neg - 1, bits)
+
+    # Assuming we would ever add them, (u)int0 types are logically integer
+    # types with only one representable value: 0
```

Comment:
I agree that a `uint0` would be 0, but I'd say it's less clear whether an `int0` is 0 or -1. With a C99 mindset, I would say it's UB since the result depends on the nonexistent sign bit.

Reply:
`int0` is strange enough of a concept that I'd say its definition would be very much open to interpretation. I'd say 0 would make more sense than -1, but as you said, either way the fact the sign bit can't be represented makes it incompatible with C99.